##
## Temi
##

#### Set up

In [12]:

def generate_batch_n_elems(iterable, n=1):
    l = len(iterable)
    for ndx in range(0, l, n):
        yield iterable[ndx:min(ndx + n, l)]

In [13]:
def generate_n_batches(lst, batch_n, len_lst = None):
    """
    Given a list, this function yields batches of an unspecified size but the number of batches is equal to `batch_n`
    E.g. generate_batch([0, 1, 2, 3, 4, 5, 6], batch_n=2) -> (0, 1, 2, 3), (4, 5, 6)
    
    Parameters:
        lst: list
        batch_n: int
            Number of batches to return
        len_lst: None or num (length of the input list)

    Yields
        `batch_n` batches of the list
    """
    import math
    # how many per batch
    if len_lst is not None:
        n_elems = math.ceil(len_lst/batch_n)
    else:
        n_elems = math.ceil(len(lst)/batch_n)
        
    for i in range(0, len(lst), n_elems):
        yield lst[i:(i + n_elems)]

In [14]:
ll = list(range(0, 500))

In [17]:
len(list(generate_batch_n_elems(ll, n=5)))

100

In [18]:
len(list(generate_n_batches(ll, 5)))

5

In [1]:
import os
import numpy as np
import importlib

In [2]:
os.getcwd()
os.chdir('/lus/grand/projects/covid-ct/imlab/users/temi/projects/TFXcan/TFPred_pipeline/test_run')

### Test the multiprocessing pool to aggregate multiple files

I use this to test my functions that help to aggregate the predictions

In [4]:
import utils
importlib.reload(utils)

<module 'utils' from '/lus/grand/projects/covid-ct/imlab/users/temi/projects/TFXcan/TFPred_pipeline/helpers/utils.py'>

In [5]:
h5_files_dir = '../prediction_folder/enformer_predictions/cistrome_AR/predictions_2023-01-24/cistrome/'
os.path.isdir(h5_files_dir)

True

Here I will read 20 motifs

In [10]:
h5_files = os.listdir(h5_files_dir)[0:20]
motifs = ['_'.join(h.split('_')[0:3]) for h in h5_files]
motifs[0:4], len(motifs)

(['chr9_66943762_66943771',
  'chr11_130862021_130862030',
  'chr8_2573263_2573272',
  'chr1_211474183_211474192'],
 20)

In [11]:
h5_predictions = [utils.read_file(m, h5_files_dir) for m in motifs]
h5_predictions =  {k: v for d in h5_predictions for k, v in d.items()}
h5_predictions

{'chr9_66943762_66943771': array([[0.08063664, 0.12737752, 0.11388403, ..., 0.00596294, 0.03120329,
         0.04742806],
        [0.13302283, 0.18834506, 0.0856443 , ..., 0.01001805, 0.06463382,
         0.08324532],
        [0.07578191, 0.10179935, 0.06314178, ..., 0.00517672, 0.02608535,
         0.0223082 ],
        ...,
        [0.04952738, 0.05461145, 0.06767036, ..., 0.00909393, 0.02509049,
         0.01374595],
        [0.06451581, 0.06840262, 0.0955685 , ..., 0.01183867, 0.03387516,
         0.0167965 ],
        [0.05021173, 0.05038195, 0.05846206, ..., 0.00364656, 0.01734603,
         0.01089448]], dtype=float32),
 'chr11_130862021_130862030': array([[0.12927958, 0.12927222, 0.10939392, ..., 0.00330115, 0.04640604,
         0.02006413],
        [0.1803764 , 0.17099352, 0.14611177, ..., 0.01003553, 0.08119576,
         0.02783836],
        [0.18431626, 0.18088901, 0.1655869 , ..., 0.00866537, 0.06991176,
         0.0299842 ],
        ...,
        [0.21393728, 0.22304513, 0.139

In [14]:
pred_mean = utils.agg_by_mean(h5_predictions)
pred_mean.shape

(20, 5314)

### Query VCFs and changing variants

Here I test how to query vcf files and create variants

In [39]:
import cyvcf2, kipoiseq
import os, sys
import numpy as np
import importlib
import subprocess

GPUs might become necessary down the line

In [40]:
subprocess.call('nvidia-smi') # just checking if you have GPUs available

Failed to initialize NVML: Insufficient Permissions


4

In [41]:
os.chdir('/lus/grand/projects/covid-ct/imlab/users/temi/projects/TFXcan/TFPred_pipeline/test_run')
os.getcwd()

'/lus/grand/projects/covid-ct/imlab/users/temi/projects/TFXcan/TFPred_pipeline/test_run'

I am going to use the enformer predict modules

In [42]:
sys.path.append('../parallel_enformer/batch_utils')
sys.path

['/lus/grand/projects/covid-ct/imlab/users/temi/projects/TFXcan/TFPred_pipeline/helpers',
 '/home/temi/miniconda3/envs/dl-tools/lib/python310.zip',
 '/home/temi/miniconda3/envs/dl-tools/lib/python3.10',
 '/home/temi/miniconda3/envs/dl-tools/lib/python3.10/lib-dynload',
 '',
 '/home/temi/miniconda3/envs/dl-tools/lib/python3.10/site-packages',
 '../parallel_enformer/batch_utils',
 '../parallel_enformer/batch_utils']

##### Here I test how manipulating reference sequences may work

Import `sequencesUtils`

In [76]:
# import utilities
# importlib.reload(utilities)

import sequencesUtils
importlib.reload(sequencesUtils)

<module 'sequencesUtils' from '/lus/grand/projects/covid-ct/imlab/users/temi/projects/TFXcan/TFPred_pipeline/test_run/../parallel_enformer/batch_utils/sequencesUtils.py'>

In [8]:
samples = ['LuCaP_173', 'LuCaP_145', 'LuCaP_70', 'LuCap_00'] # LuCaP_000 does not exist

In [9]:
path_to_vcf= '../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz'
if os.path.isfile(path_to_vcf):
    myvcf = cyvcf2.cyvcf2.VCF(path_to_vcf, samples=samples)
else: print('vcf file not found')

/tmp/ipykernel_166557/219678836.py:3: UserWarning: not all requested samples found in VCF
  myvcf = cyvcf2.cyvcf2.VCF(path_to_vcf, samples=samples)
[W::hts_idx_load3] The index file is older than the data file: ../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz.tbi


In [10]:
path_to_fasta = '../../../../../../data/hg_sequences/hg38/Homo_sapiens_assembly38.fasta'
if os.path.isfile(path_to_fasta):   
    fasta_object = sequencesUtils.FastaStringExtractor(path_to_fasta)
else: print('Fasta file not found')

You can get the samples read by the vcf parser this way

In [11]:
myvcf.samples

['LuCaP_145', 'LuCaP_173', 'LuCaP_70']

And given a region, you can extract variants this way

In [45]:
short_query_region = 'chr17:71028804-71028813'
for v in myvcf(short_query_region):
    print(v.CHROM, v.POS, v.genotypes, v.gt_bases)

Extract that region from the vcf file

In [46]:
short_query_region = short_query_region.replace(':', '_').replace('-', '_')
short_query_region # I need this format

'chr17_71028804_71028813'

In [58]:
ref_seq = sequencesUtils.extract_reference_sequence(region=short_query_region, fasta_func=fasta_object, resize_length=None, print_sequence=False, resize_for_enformer=True)
ref_seq, ref_seq['sequence'].shape

({'sequence': array([[[0., 1., 0., 0.],
          [1., 0., 0., 0.],
          [1., 0., 0., 0.],
          ...,
          [0., 1., 0., 0.],
          [0., 0., 0., 1.],
          [1., 0., 0., 0.]]], dtype=float32),
  'interval_object': Interval(chrom='chr17', start=70832201, end=71225417, name='', strand='.', ...)},
 (1, 393216, 4))

Find variants belonging to that region

In [68]:
variants_in_samples = sequencesUtils.find_variants_in_vcf_file(cyvcf2_object=myvcf, interval_object=ref_seq['interval_object'], samples=samples)
variants_in_samples

{'chr': 'chr17',
 'positions': (70832442,
  70834957,
  70835259,
  70835380,
  70835568,
  70835922,
  70835941,
  70835954,
  70836131,
  70836141,
  70836243,
  70836303,
  70837185,
  70838228,
  70838239,
  70838244,
  70838499,
  70839073,
  70839541,
  70839554,
  70841533,
  70841722,
  70842960,
  70844321,
  70844792,
  70846241,
  70846922,
  70848157,
  70848894,
  70848944,
  70850471,
  70850545,
  70850638,
  70850845,
  70851323,
  70851331,
  70851546,
  70851612,
  70852480,
  70852604,
  70852657,
  70852758,
  70852787,
  70852793,
  70852795,
  70852796,
  70852800,
  70853574,
  70853673,
  70853783,
  70853888,
  70853966,
  70854167,
  70854361,
  70855030,
  70855318,
  70856773,
  70856972,
  70858179,
  70859131,
  70859306,
  70859323,
  70859689,
  70859753,
  70859798,
  70859946,
  70861032,
  70861105,
  70861303,
  70861511,
  70861939,
  70861976,
  70862050,
  70862072,
  70862243,
  70862544,
  70862614,
  70862919,
  70863998,
  70863999,
  70864153

`chr` is the chromosome

`positions` are the positions

and the rest are the individuals and their sequences at those positions

I need to create a mapping dictionary that turns these sequences into one-hot encoded numpy arrays

In [73]:
variants_map = sequencesUtils.create_mapping_dictionary(variants_in_samples, ref_seq['interval_object'].start, sequence_source='personalized')
variants_map

{'positions': (241,
  2756,
  3058,
  3179,
  3367,
  3721,
  3740,
  3753,
  3930,
  3940,
  4042,
  4102,
  4984,
  6027,
  6038,
  6043,
  6298,
  6872,
  7340,
  7353,
  9332,
  9521,
  10759,
  12120,
  12591,
  14040,
  14721,
  15956,
  16693,
  16743,
  18270,
  18344,
  18437,
  18644,
  19122,
  19130,
  19345,
  19411,
  20279,
  20403,
  20456,
  20557,
  20586,
  20592,
  20594,
  20595,
  20599,
  21373,
  21472,
  21582,
  21687,
  21765,
  21966,
  22160,
  22829,
  23117,
  24572,
  24771,
  25978,
  26930,
  27105,
  27122,
  27488,
  27552,
  27597,
  27745,
  28831,
  28904,
  29102,
  29310,
  29738,
  29775,
  29849,
  29871,
  30042,
  30343,
  30413,
  30718,
  31797,
  31798,
  31952,
  32213,
  32308,
  32594,
  32703,
  32846,
  32871,
  33188,
  33775,
  34349,
  34350,
  34367,
  34369,
  34371,
  34447,
  34462,
  35136,
  35392,
  35616,
  36320,
  36624,
  36835,
  37928,
  38274,
  38302,
  38495,
  39906,
  40257,
  40422,
  41370,
  41411,
  41536,
  

In [74]:
np.max(variants_map['positions'])

393216

In [92]:

import sequencesUtils
importlib.reload(sequencesUtils)

<module 'sequencesUtils' from '/lus/grand/projects/covid-ct/imlab/users/temi/projects/TFXcan/TFPred_pipeline/test_run/../parallel_enformer/batch_utils/sequencesUtils.py'>

In [93]:
variants_sequences = sequencesUtils.replace_variants_in_reference_sequence(ref_seq['sequence'], variants_map)
variants_sequences

(array([0., 0., 0., 1.], dtype=float32), array([0., 0., 1., 0.], dtype=float32), array([0., 0., 0., 1.], dtype=float32), array([0., 0., 0., 1.], dtype=float32), array([1., 0., 0., 0.], dtype=float32))
393215 in indices
[1. 0. 0. 0.]
393216
[393141, 393145, 393154, 393176, 393215]
[array([0., 0., 0., 1.], dtype=float32), array([0., 0., 0., 1.], dtype=float32), array([0., 0., 1., 0.], dtype=float32), array([0., 0., 0., 1.], dtype=float32), array([0., 0., 0., 1.], dtype=float32)]
[INFO] Encountered an index error for LuCaP_173. Index 393216 was found. Correcting and removed 393216 & [1. 0. 0. 0.]
(array([0., 0., 0., 1.], dtype=float32), array([0., 0., 1., 0.], dtype=float32), array([0., 0., 0., 1.], dtype=float32), array([0., 0., 0., 1.], dtype=float32), array([1., 0., 0., 0.], dtype=float32))
393215 in indices
[1. 0. 0. 0.]
393216
[393141, 393145, 393154, 393176, 393215]
[array([0., 0., 0., 1.], dtype=float32), array([0., 0., 0., 1.], dtype=float32), array([0., 0., 1., 0.], dtype=float32

{'LuCaP_173': {'haplotype1': array([[[0., 1., 0., 0.],
          [1., 0., 0., 0.],
          [1., 0., 0., 0.],
          ...,
          [0., 1., 0., 0.],
          [0., 0., 0., 1.],
          [0., 0., 0., 1.]]], dtype=float32),
  'haplotype2': array([[[0., 1., 0., 0.],
          [1., 0., 0., 0.],
          [1., 0., 0., 0.],
          ...,
          [0., 1., 0., 0.],
          [0., 0., 0., 1.],
          [0., 0., 0., 1.]]], dtype=float32)},
 'LuCaP_145': {'haplotype1': array([[[0., 1., 0., 0.],
          [1., 0., 0., 0.],
          [1., 0., 0., 0.],
          ...,
          [0., 1., 0., 0.],
          [0., 0., 0., 1.],
          [0., 0., 0., 1.]]], dtype=float32),
  'haplotype2': array([[[0., 1., 0., 0.],
          [1., 0., 0., 0.],
          [1., 0., 0., 0.],
          ...,
          [0., 1., 0., 0.],
          [0., 0., 0., 1.],
          [0., 0., 0., 1.]]], dtype=float32)},
 'LuCaP_70': {'haplotype1': array([[[0., 1., 0., 0.],
          [1., 0., 0., 0.],
          [1., 0., 0., 0.],
  

In [66]:
er = (4,3,2,1)
er.index(1)

3

Now, I have the indices of the positions (as opposed to coordinates), and the individuals have been converted to np arrays 

In [107]:
# Here I define a function to reverse_one_hot_encode a numpy array

def reverse_encode_into_sequences(input_array, delim_indices):

    import copy 
    # never change this whatsoever
    A = np.array([1,0,0,0], dtype=np.float32)
    C = np.array([0,1,0,0], dtype=np.float32)
    G = np.array([0,0,1,0], dtype=np.float32)
    T = np.array([0,0,0,1], dtype=np.float32)
    seq_dict = {'A': A, 'C': C, 'G': G, 'T': T}

    keys = list(seq_dict.keys())
    values = list(seq_dict.values())
    values = [ar.tolist() for ar in values]

    if len(input_array.shape) != 3:
        raise Exception('Input array should have 3 dimensions')

    seq_list = []
    for i, na in enumerate(range(input_array.shape[1])):
        ts = input_array[:,i,:].tolist()[0]
        # look up in the dictionary
        id = values.index(ts)
        seq_list.append(keys[id])

    seq = [seq_list[i] if i in delim_indices else seq_list[i].lower() for i in range(len(seq_list)) ] #

    return(''.join(seq))

In [108]:
reverse_encode_into_sequences(ref_seq['sequence'], variants_map['positions'])

'caagatccTatctcaaaaaaaaaaaaAaaa'

I go through each individual's haplotypes variants sequence, reverse encode the already encoded sequences, and see that they have been changed appropriately in the reference sequence

Variants are in upper case

In [109]:
print(f'Reference sequence: {reverse_encode_into_sequences(ref_seq["sequence"], variants_map["positions"])}')

for sample, dt in variants_sequences.items():
    for haplotype, hap_sequences in dt.items():
        print(f'{sample}-{haplotype}: {reverse_encode_into_sequences(hap_sequences, variants_map["positions"])}')

Reference sequence: caagatccTatctcaaaaaaaaaaaaAaaa
LuCaP_173-haplotype1: caagatccCatctcaaaaaaaaaaaaAaaa
LuCaP_173-haplotype2: caagatccCatctcaaaaaaaaaaaaAaaa
LuCaP_145-haplotype1: caagatccCatctcaaaaaaaaaaaaAaaa
LuCaP_145-haplotype2: caagatccCatctcaaaaaaaaaaaaAaaa
LuCaP_70-haplotype1: caagatccCatctcaaaaaaaaaaaaAaaa
LuCaP_70-haplotype2: caagatccCatctcaaaaaaaaaaaaAaaa


### Create encoded input sequences for ENFORMER

-: Handle edge cases
1. the region does not have a variant in the vcf file
2. a sample is not the vcf file
3. the sequences from the reference genome are `NNNNNNNNNN...`

In [110]:
query_list = ['chr1:103268936-103268945', 'chr1:150333985-150333994', 'chr12:45119102-45119111', 'chr8:146200116-146200125']
query_list = ['chr1_103268936_103268945', 'chr1_150333985_150333994', 'chr12_45119102_45119111', 'chr8_146200116_146200125', 'chr1_20_50']

The regions need to be in a certain format

In [111]:
batch_regions = [{'query': q, 'logtype': 'y'} for q in query_list]
batch_regions

[{'query': 'chr1_103268936_103268945', 'logtype': 'y'},
 {'query': 'chr1_150333985_150333994', 'logtype': 'y'},
 {'query': 'chr12_45119102_45119111', 'logtype': 'y'},
 {'query': 'chr8_146200116_146200125', 'logtype': 'y'},
 {'query': 'chr1_20_50', 'logtype': 'y'}]

In [126]:
samples_enformer_inputs = [sequencesUtils.create_input_for_enformer(region_details=loci, samples=samples, 
    path_to_vcf=path_to_vcf, 
    fasta_func=fasta_object, hap_type = 'both', 
    resize_for_enformer=False, resize_length=None, 
    write_log=None, sequence_source='personalized') for loci in batch_regions]

samples_enformer_inputs

[W::hts_idx_load3] The index file is older than the data file: ../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz.tbi
/lus/grand/projects/covid-ct/imlab/users/temi/projects/TFXcan/TFPred_pipeline/test_run/../parallel_enformer/batch_utils/sequencesUtils.py:314: UserWarning: not all requested samples found in VCF
  vcf_chr = cyvcf2.cyvcf2.VCF(path_to_vcf, samples=samples)
[W::hts_idx_load3] The index file is older than the data file: ../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz.tbi
[W::hts_idx_load3] The index file is older than the data file: ../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz.tbi
[W::hts_idx_load3] The index file is older than the data file: ../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz.tbi
[W::hts_idx_load3] The index file is older than the data file: ../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz.tbi
[W::hts_idx_load3] The index file is older than the data file: ../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz.tbi


[{'sequence': {'haplotype1': array([[[0., 1., 0., 0.],
           [1., 0., 0., 0.],
           [0., 0., 1., 0.],
           [0., 1., 0., 0.],
           [0., 1., 0., 0.],
           [0., 1., 0., 0.],
           [0., 1., 0., 0.],
           [0., 0., 0., 1.],
           [0., 1., 0., 0.]]], dtype=float32),
   'haplotype2': array([[[0., 1., 0., 0.],
           [1., 0., 0., 0.],
           [0., 0., 1., 0.],
           [0., 1., 0., 0.],
           [0., 1., 0., 0.],
           [0., 1., 0., 0.],
           [0., 1., 0., 0.],
           [0., 0., 0., 1.],
           [0., 1., 0., 0.]]], dtype=float32)},
  'metadata': {'sequence_source': 'ref',
   'region': 'chr1_103268936_103268945',
   'logtype': 'y',
   'samples': ['LuCaP_173', 'LuCaP_145', 'LuCaP_70', 'LuCap_00']}},
 {'sequence': {'LuCaP_173': {'haplotype1': array([[[1., 0., 0., 0.],
            [1., 0., 0., 0.],
            [1., 0., 0., 0.],
            [0., 0., 0., 1.],
            [1., 0., 0., 0.],
            [0., 0., 0., 1.],
            [

In [119]:
len(samples_enformer_inputs)

5

In [124]:
samples_enformer_inputs[2]

{'sequence': {'haplotype1': array([[[0., 0., 0., 1.],
          [0., 1., 0., 0.],
          [0., 0., 0., 1.],
          [0., 1., 0., 0.],
          [0., 0., 0., 1.],
          [1., 0., 0., 0.],
          [1., 0., 0., 0.],
          [0., 0., 0., 1.],
          [0., 0., 0., 1.]]], dtype=float32),
  'haplotype2': array([[[0., 0., 0., 1.],
          [0., 1., 0., 0.],
          [0., 0., 0., 1.],
          [0., 1., 0., 0.],
          [0., 0., 0., 1.],
          [1., 0., 0., 0.],
          [1., 0., 0., 0.],
          [0., 0., 0., 1.],
          [0., 0., 0., 1.]]], dtype=float32)},
 'metadata': {'sequence_source': 'ref',
  'region': 'chr12_45119102_45119111',
  'logtype': 'y',
  'samples': ['LuCaP_173', 'LuCaP_145', 'LuCaP_70', 'LuCap_00']}}

In [90]:
ref_seq = sequencesUtils.extract_reference_sequence(region=query_list[1], fasta_func=fasta_object, resize_length=50, resize_for_enformer=True)
ref_seq, ref_seq['sequence'].shape

extracting reference sequence
finished extracting reference sequence of length 393216


({'sequence': array([[[1., 0., 0., 0.],
          [1., 0., 0., 0.],
          [0., 0., 0., 1.],
          ...,
          [0., 0., 1., 0.],
          [1., 0., 0., 0.],
          [1., 0., 0., 0.]]], dtype=float32),
  'interval_object': Interval(chrom='chr1', start=150137382, end=150530598, name='', strand='.', ...)},
 (1, 393216, 4))

In [ ]:
variants_in_samples = find_variants_in_vcf_file(cyvcf2_object=myvcf, interval_object=ref_seq['interval_object'], samples=samples[0:3])

In [ ]:
#  look at
variants_in_samples['positions'] ; variants_in_samples['LuCaP_173']

In [20]:
samples_mapping_dictionary = create_mapping_dictionary(variants_in_samples, ref_seq['interval_object'].start, haplotype='both')

In [22]:
samples_mapping_dictionary.keys()

dict_keys(['positions', 'LuCaP_173', 'LuCaP_145', 'LuCaP_70'])

In [ ]:
replace_variants_in_reference_sequence(ref_seq['sequence'], samples_mapping_dictionary)

In [14]:
def create_input_for_enformer(region_details, samples, path_to_vcf, fasta_func, hap_type = 'both', vcf_func=None, resize_for_enformer=True, resize_length=None, write_log=None, sequence_source=None):
    '''
    Given a region in the genome, a reference genome (fasta) and a VCF file, create an individual's sequence for that region

    Parameters:
        query_sequences: str
            The query, perhaps, reference sequence
        mapping_dict: a numpy array or list
            List of the regions that should be modified

    Returns:
        A new sequence with all the changes applied.
    '''

    import numpy as np
    import cyvcf2

    #print(region_details)
    region = region_details['query']
    logtype = region_details['logtype']

    if sequence_source == 'random':
        return({'sequence': {'haplotype0': one_hot_encode(generate_random_sequence_inputs())}, 'metadata': {'sequence_source':'random', 'region':region, 'logtype': logtype}})
    else:
        reference_sequence = extract_reference_sequence(region, fasta_func, resize_for_enformer)
        if np.all(reference_sequence['sequence'] == 0.25): # check if all the sequence are "NNNNNNNNNNN..."
            if write_log['error']:
                err_msg = f'[INPUT ERROR] {region} is invalid; all nucleotides are N.'
                MEMORY_ERROR_FILE = f"{log_dir}/error_details.log"
                setup_logger('error_log', MEMORY_ERROR_FILE)
                logger(err_msg, 'error', 'run_error')
            return(None)

        else:
            if sequence_source == 'reference':
                return({'sequence': {'haplotype0': reference_sequence['sequence']}, 'metadata': {'sequence_source':'ref', 'region':region, 'logtype': logtype}})
            elif sequence_source == 'personalized':
                # which vcf file
                vcf_chr = cyvcf2.cyvcf2.VCF(path_to_vcf, samples=samples)
                samples_variants = find_variants_in_vcf_file(cyvcf2_object=vcf_chr, interval_object=reference_sequence['interval_object'], samples=samples)
                if samples_variants: # go on and change the variants by position
                    samples_mapping_dictionary = create_mapping_dictionary(samples_variants, reference_sequence['interval_object'].start, haplotype='both')
                    samples_variants_encoded = replace_variants_in_reference_sequence(reference_sequence['sequence'], samples_mapping_dictionary)
                    return({'sequence': samples_variants_encoded, 'metadata': {'sequence_source':'var', 'region':region, 'logtype': logtype}})
                else: # return the reference and no need for having the same reference genome for all samples since one is enough 
                    return({'sequence':{'haplotype1': reference_sequence['sequence'], 'haplotype2': reference_sequence['sequence']}, 'metadata': {'sequence_source':'ref', 'region':region, 'logtype': logtype, 'samples': samples}})

[{'query': 'chr1_103268936_103268945', 'logtype': 'y'},
 {'query': 'chr1_150333985_150333994', 'logtype': 'y'},
 {'query': 'chr12_45119102_45119111', 'logtype': 'y'},
 {'query': 'chr8_146200116_146200125', 'logtype': 'y'}]

In [16]:
samples

['LuCaP_173', 'LuCaP_145', 'LuCaP_70']

In [131]:
samples_enformer_inputs = create_input_for_enformer(region_details=batch_regions[1], samples=samples, path_to_vcf=path_to_vcf, fasta_func=fasta_object, hap_type = 'both', resize_for_enformer=True, resize_length=None, write_log=None, sequence_source='reference')

samples_enformer_inputs

{'sequence': {'haplotype0': array([[[1., 0., 0., 0.],
          [1., 0., 0., 0.],
          [0., 0., 0., 1.],
          ...,
          [0., 0., 1., 0.],
          [1., 0., 0., 0.],
          [1., 0., 0., 0.]]], dtype=float32)},
 'metadata': {'sequence_source': 'ref',
  'region': 'chr1_150333985_150333994',
  'logtype': 'y'}}

In [19]:
import tensorflow as tf

In [20]:
output_dir = './test_output'
if not os.path.exists(output_dir): os.makedirs(output_dir)

global write_log
write_log = {'memory':False, 'error':False, 'time':False, 'cache':False}

global model
model = tf.saved_model.load('../../../../../../data/enformer/raw').model
fasta_object = utilities.FastaStringExtractor('../../../../../../data/hg_sequences/hg38/Homo_sapiens_assembly38.fasta')

2023-02-03 05:08:52.860049: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-02-03 05:08:54.467357: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1532] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 38417 MB memory:  -> device: 0, name: NVIDIA A100-SXM4-40GB, pci bus id: 0000:07:00.0, compute capability: 8.0
2023-02-03 05:08:54.469323: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1532] Created device /job:localhost/replica:0/task:0/device:GPU:1 with 38417 MB memory:  -> device: 1, name: NVIDIA A100-SXM4-40GB, pci bus id: 0000:46:00.0, compute capability: 8.0
2023-02-03 05:08:54.470568: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1532] Created device /job:localhost/replica:0/task:0/devi

In [21]:
batch_regions, samples

([{'query': 'chr1_103268936_103268945', 'logtype': 'y'},
  {'query': 'chr1_150333985_150333994', 'logtype': 'y'},
  {'query': 'chr12_45119102_45119111', 'logtype': 'y'},
  {'query': 'chr8_146200116_146200125', 'logtype': 'y'}],
 ['LuCaP_173', 'LuCaP_145', 'LuCaP_70'])

In [29]:
enformer_predict_on_batch(batch_regions=batch_regions, samples=samples, path_to_vcf=path_to_vcf, batch_num=1, output_dir=output_dir, predictions_log_file='./test_output/three_people.csv')

[W::hts_idx_load3] The index file is older than the data file: ../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz.tbi
[W::hts_idx_load3] The index file is older than the data file: ../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz.tbi


Writing log file
logfile found


[W::hts_idx_load3] The index file is older than the data file: ../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz.tbi
[W::hts_idx_load3] The index file is older than the data file: ../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz.tbi


Writing log file
logfile found


[W::hts_idx_load3] The index file is older than the data file: ../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz.tbi
[W::hts_idx_load3] The index file is older than the data file: ../../genotypes/prj6_genotypes/merged_phased_SNPs.vcf.gz.tbi


Writing log file
logfile found


In [22]:
import itertools

In [28]:
# check
def check_queries(sample, queries, output_dir, logfile, sequence_source):
    """
    Check whether a given region, for an individual has been predicted and logged.

    Parameters:
        sample: str 
            The name/id of an individual
        queries: A batch
            A region in the genome in the form `chr_start_end`.
        output_dir: str (path)
            The folder where the predictions should have been logged. 
        logfile: pd.DataFrame 
            A dataframe of a log file or `None` if the log file does not exist. 
    
    Returns: dict
        'query': the query region if it has not been logged or predictions don't exist
        'logtype': whether it should be logged if it has not been logged

    If predictions exist and the query has been logged, this function returns None.
    """
    import pandas as pd
    import os
    import numpy as np

    if isinstance(logfile, type(None)):
        output = [{'query':query, 'logtype':'y'} for query in queries]
        return(output)

    elif isinstance(logfile, pd.DataFrame):# should have read it>

        id_logfile = logfile.loc[logfile['sample'] == sample, : ]
        # check if the file is saved
        if sequence_source == 'personalized': # prediction must be present in two folders
            queries_saved_h1 = [str(f'{output_dir}/{sample}/haplotype1/{query}_predictions.h5') for query in queries]
            queries_saved_h2 = [str(f'{output_dir}/{sample}/haplotype2/{query}_predictions.h5') for query in queries]
            queries_saved = np.array([os.path.isfile(q) for q in queries_saved_h1]) * np.array([os.path.isfile(q) for q in queries_saved_h2])
        elif sequence_source == 'reference':
            queries_saved = [str(f'{output_dir}/{sample}/haplotype0/{query}_predictions.h5') for query in queries]
            queries_saved = np.array([os.path.isfile(q) for q in queries_saved])
        # check if the file is in the logfile
        query_logged = np.array([(query in id_logfile.motif.values) for query in queries])
        # 
        if not query_logged.shape[0] == queries_saved.shape[0]:
            raise Exception("Lengths of queries logged and saved conditions are not the same")
        queries_condition = queries_saved * query_logged
        queries_condition = queries_condition.tolist()
        # 
        output = [{'query':queries[i], 'logtype':'n'} if qc is True else {'query':queries[i], 'logtype':'y'} for i, qc in enumerate(queries_condition)]

        return(output)


In [29]:
query_list = ['chr1_103268936_103268945', 'chr1_150333985_150333994', 'chr12_45119102_45119111', 'chr8_146200116_146200125']

In [30]:
import pandas as pd

pt = pd.read_csv('./test_output/three_people.csv')
pt

motif     sample     status sequence_source
0   chr1_103268936_103268945  LuCaP_173  completed             ref
1   chr1_103268936_103268945  LuCaP_145  completed             ref
2   chr1_103268936_103268945   LuCaP_70  completed             ref
3   chr1_150333985_150333994  LuCaP_173  completed             ref
4   chr1_150333985_150333994  LuCaP_145  completed             ref
5   chr1_150333985_150333994   LuCaP_70  completed             ref
6    chr12_45119102_45119111  LuCaP_173  completed             ref
7    chr12_45119102_45119111  LuCaP_145  completed             ref
8    chr12_45119102_45119111   LuCaP_70  completed             ref
9   chr1_103268936_103268945  LuCaP_173  completed             var
10  chr1_103268936_103268945  LuCaP_145  completed             var
11  chr1_103268936_103268945   LuCaP_70  completed             var
12  chr1_150333985_150333994  LuCaP_173  completed             var
13  chr1_150333985_150333994  LuCaP_145  completed             var
14  chr1_150333985_150333994   LuCaP_70  completed             var
15   chr12_45119102_45119111  LuCaP_173  completed             var
16   chr12_45119102_45119111  LuCaP_145  completed             var
17   chr12_45119102_45119111   LuCaP_70  completed             var

In [31]:
check_these = list(itertools.product(samples, [query_list]))
check_these

[('LuCaP_173',
  ['chr1_103268936_103268945',
   'chr1_150333985_150333994',
   'chr12_45119102_45119111',
   'chr8_146200116_146200125']),
 ('LuCaP_145',
  ['chr1_103268936_103268945',
   'chr1_150333985_150333994',
   'chr12_45119102_45119111',
   'chr8_146200116_146200125']),
 ('LuCaP_70',
  ['chr1_103268936_103268945',
   'chr1_150333985_150333994',
   'chr12_45119102_45119111',
   'chr8_146200116_146200125'])]

In [32]:
checked = [check_queries(sample=cq[0], queries=cq[1], output_dir=output_dir, logfile=pt, sequence_source='personalized') for cq in check_these]
checked

[[{'query': 'chr1_103268936_103268945', 'logtype': 'n'},
  {'query': 'chr1_150333985_150333994', 'logtype': 'n'},
  {'query': 'chr12_45119102_45119111', 'logtype': 'n'},
  {'query': 'chr8_146200116_146200125', 'logtype': 'y'}],
 [{'query': 'chr1_103268936_103268945', 'logtype': 'n'},
  {'query': 'chr1_150333985_150333994', 'logtype': 'n'},
  {'query': 'chr12_45119102_45119111', 'logtype': 'n'},
  {'query': 'chr8_146200116_146200125', 'logtype': 'y'}],
 [{'query': 'chr1_103268936_103268945', 'logtype': 'n'},
  {'query': 'chr1_150333985_150333994', 'logtype': 'n'},
  {'query': 'chr12_45119102_45119111', 'logtype': 'n'},
  {'query': 'chr8_146200116_146200125', 'logtype': 'y'}]]

In [45]:
ff = [f for f in checked for f in f]
ff

[{'query': 'chr1_103268936_103268945', 'logtype': 'n'},
 {'query': 'chr1_150333985_150333994', 'logtype': 'n'},
 {'query': 'chr12_45119102_45119111', 'logtype': 'n'},
 {'query': 'chr8_146200116_146200125', 'logtype': 'y'},
 {'query': 'chr1_103268936_103268945', 'logtype': 'n'},
 {'query': 'chr1_150333985_150333994', 'logtype': 'n'},
 {'query': 'chr12_45119102_45119111', 'logtype': 'n'},
 {'query': 'chr8_146200116_146200125', 'logtype': 'y'},
 {'query': 'chr1_103268936_103268945', 'logtype': 'n'},
 {'query': 'chr1_150333985_150333994', 'logtype': 'n'},
 {'query': 'chr12_45119102_45119111', 'logtype': 'n'},
 {'query': 'chr8_146200116_146200125', 'logtype': 'y'}]

In [47]:
list({v['query']: v for v in ff}.values())

[{'query': 'chr1_103268936_103268945', 'logtype': 'n'},
 {'query': 'chr1_150333985_150333994', 'logtype': 'n'},
 {'query': 'chr12_45119102_45119111', 'logtype': 'n'},
 {'query': 'chr8_146200116_146200125', 'logtype': 'y'}]

In [43]:
{k: v for m in ff for k, v in m.items()} # {k: v for d in ini_dict for k, v in d.items()}

{'query': 'chr8_146200116_146200125', 'logtype': 'y'}

In [50]:
ind_var = utilities.find_variants_in_vcf_file(myvcf, ref_seq['interval_object'])
ind_var[0:5], len(ind_var)

([['chr1', 103072672, [0, 0], ['T', 'T']],
  ['chr1', 103073435, [0, 0], ['A', 'A']],
  ['chr1', 103073451, [0, 0], ['A', 'A']],
  ['chr1', 103074415, [0, 0], ['C', 'C']],
  ['chr1', 103074421, [0, 0], ['G', 'G']]],
 1602)

In [51]:
map_dict = utilities.create_mapping_dictionary(ind_var, ref_seq['interval_object'].start, haplotype='both')

# looking inside
len(map_dict) , map_dict.keys()

(2, dict_keys(['haplotype1', 'haplotype2']))

In [56]:
var_seq = utilities.replace_variants_in_reference_sequence(ref_seq['sequence'], map_dict)
len(var_seq)

(1, 393216, 4)
(1, 393216, 4)


2

In [32]:
ref_seq['sequence'].shape

(1, 393216, 4)

In [53]:
var_seq['haplotype1'][:, 4371, :] , var_seq['haplotype2'][:, 4371, :] #>> ['chr1', 103076704, [0, 1], ['A', 'G']]

(array([[1., 0., 0., 0.]], dtype=float32),
 array([[0., 0., 1., 0.]], dtype=float32))

In [54]:
bool_hap = var_seq['haplotype1'] == var_seq['haplotype2']
bool_hap # some of these rows will be false because there are variants
#np.all(bool_hap == True, axis=0)

array([[[ True,  True,  True,  True],
        [ True,  True,  True,  True],
        [ True,  True,  True,  True],
        ...,
        [ True,  True,  True,  True],
        [ True,  True,  True,  True],
        [ True,  True,  True,  True]]])

#### Verifying

I can look at the true indices where there are variants, and compare that to the indices that are different between haplotypes 1 and 2 and see that they are the same. If they are the same, then my approach is working correctly.

In [15]:
def get_different_indices(ind_var, st):
    out = [(ind_var[i][1] - st) for i, _ in enumerate(ind_var) if ind_var[i][2][0] != ind_var[i][2][1]]
    return(out)

In [16]:
true_changes = get_different_indices(ind_var, ref_seq['interval_object'].start)
true_changes.sort(), len(true_changes)

(None, 183)

In [17]:
#made_changes = list(set(np.where(bool_hap == False)[1].tolist())) # ~a.any(axis=1)
made_changes = list(set(np.where(~bool_hap.any(axis=0))[0].tolist())) 
made_changes.sort(), len(made_changes)

(None, 183)

In [18]:
made_changes == true_changes

True

### Now I can put together all the code

Timing this function on 4 individuals

In [27]:
import time

tic = time.perf_counter()
out = [utilities.create_input_for_enformer(region_details=b, fasta_func=fasta_object, hap_type = 'both', vcf_func=myvcf, resize_for_enformer=True, resize_length=None, write_log=utilities.write_log, sequence_source='personalized') for b in batch_region]
toc = time.perf_counter()

(toc - tic) / 4

{'query': 'chr1_103268936_103268945', 'logtype': 'n'}
{'query': 'chr1_150333985_150333994', 'logtype': 'n'}
{'query': 'chr12_45119102_45119111', 'logtype': 'n'}
{'query': 'chr8_146200116_146200125', 'logtype': 'n'}


0.018893439235398546

### Coupling this with Enformer predict on batch

I think that having a batch of inputs (N, 393216, 4) on which to carry out inference is not necessary. Any `N` above ~ 20 results in GPU memory problems. It is better to use a for loop for all the batches. 

In [14]:
import tensorflow as tf
tf.config.experimental.list_physical_devices()

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'),
 PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU'),
 PhysicalDevice(name='/physical_device:GPU:1', device_type='GPU'),
 PhysicalDevice(name='/physical_device:GPU:2', device_type='GPU'),
 PhysicalDevice(name='/physical_device:GPU:3', device_type='GPU')]

In [15]:
enformer_model = tf.saved_model.load('../../../../../../data/enformer/raw').model

# '../../../../../../data/hg_sequences/hg38/Homo_sapiens_assembly38.fasta'

2023-02-02 22:26:19.498934: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-02-02 22:26:21.118515: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1532] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 38419 MB memory:  -> device: 0, name: NVIDIA A100-SXM4-40GB, pci bus id: 0000:07:00.0, compute capability: 8.0
2023-02-02 22:26:21.120519: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1532] Created device /job:localhost/replica:0/task:0/device:GPU:1 with 38419 MB memory:  -> device: 1, name: NVIDIA A100-SXM4-40GB, pci bus id: 0000:46:00.0, compute capability: 8.0
2023-02-02 22:26:21.121774: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1532] Created device /job:localhost/replica:0/task:0/devi

In [46]:
enformer_model.predict_on_batch(ref_seq['sequence'][np.newaxis])

2023-02-01 05:40:42.846267: I tensorflow/stream_executor/cuda/cuda_blas.cc:1786] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once.
2023-02-01 05:40:43.238039: I tensorflow/stream_executor/cuda/cuda_dnn.cc:384] Loaded cuDNN version 8100


{'human': <tf.Tensor: shape=(1, 896, 5313), dtype=float32, numpy=
 array([[[0.0448859 , 0.0338286 , 0.02249384, ..., 0.00300077,
          0.00844322, 0.0054747 ],
         [0.04086443, 0.03038296, 0.02248631, ..., 0.00447055,
          0.01325294, 0.00937136],
         [0.04369152, 0.03036681, 0.02454022, ..., 0.00513172,
          0.00897354, 0.00762524],
         ...,
         [0.02706456, 0.02111664, 0.02066585, ..., 0.00262824,
          0.00595117, 0.00492326],
         [0.03640065, 0.02561291, 0.03140229, ..., 0.00306219,
          0.00718402, 0.00764818],
         [0.03207001, 0.02182599, 0.02929399, ..., 0.00237293,
          0.00591444, 0.00633624]]], dtype=float32)>,
 'mouse': <tf.Tensor: shape=(1, 896, 1643), dtype=float32, numpy=
 array([[[0.04485861, 0.05987871, 0.0471715 , ..., 0.32980704,
          0.5065377 , 0.66570425],
         [0.04059008, 0.05534767, 0.04336205, ..., 0.2931977 ,
          0.45665944, 0.59970886],
         [0.04427745, 0.06650081, 0.06256185, ..., 

In [103]:
def enformer_predict_on_sequence(model, prepared_region):
    try:
        prediction_output = {}
        for haplotype, sequence_encoding in prepared_region['sequence'].items():
            prediction = model.predict_on_batch(sequence_encoding)['human'].numpy()[: ,range(448 - 8, (448 + 8 + 1)), : ]
            print(f'{haplotype}: {prediction.squeeze().shape}')
            prediction_output[haplotype] = prediction.squeeze()
        # delete `sequence` from the dictionary and return it
        del prepared_region['sequence']
        return( prediction_output, prepared_region )
    except ValueError:
        print(f"[ERROR] {prepared_region['region']}")
        return None
    

In [104]:
def save_batch_h5_prediction(batch_predictions, metadata, output_dir, sample):

    import h5py
    import os

    #output = []

    for key, values in batch_predictions.items():
        houtput = os.path.join(output_dir, sample, key)
        if not os.path.exists(houtput): os.makedirs(houtput)
        for i in range(0, values.shape[0]):
            region = metadata['region']
            h5save = str(f'{houtput}/{region}_predictions.h5')
            with h5py.File(h5save, 'w') as hf:
                hf.create_dataset(region, data=values[i, :])
    
    if metadata['logtype'] == 'y':
        output = [metadata['region'], sample, 'completed', metadata['sequence_source']]
    else:
        output = None

    return(output)

In [105]:
output_dir = './test_output'
if not os.path.exists(output_dir): os.makedirs(output_dir)

Testing

In [115]:
myregions = [utilities.create_input_for_enformer(region_details=b, fasta_func=fasta_object, hap_type = 'both', vcf_func=myvcf, resize_for_enformer=True, resize_length=None, write_log=utilities.write_log, sequence_source='personalized') for b in batch_region]

myregions

[{'sequence': {'haplotype1': array([[[1., 0., 0., 0.],
           [0., 0., 0., 1.],
           [1., 0., 0., 0.],
           ...,
           [1., 0., 0., 0.],
           [0., 0., 1., 0.],
           [0., 0., 0., 1.]]], dtype=float32),
   'haplotype2': array([[[1., 0., 0., 0.],
           [0., 0., 0., 1.],
           [1., 0., 0., 0.],
           ...,
           [1., 0., 0., 0.],
           [0., 0., 1., 0.],
           [0., 0., 0., 1.]]], dtype=float32)},
  'sequence_source': 'var',
  'region': 'chr1_103268936_103268945',
  'logtype': 'y'},
 {'sequence': {'haplotype1': array([[[1., 0., 0., 0.],
           [1., 0., 0., 0.],
           [0., 0., 0., 1.],
           ...,
           [0., 0., 1., 0.],
           [1., 0., 0., 0.],
           [1., 0., 0., 0.]]], dtype=float32),
   'haplotype2': array([[[1., 0., 0., 0.],
           [1., 0., 0., 0.],
           [0., 0., 0., 1.],
           ...,
           [0., 0., 1., 0.],
           [1., 0., 0., 0.],
           [1., 0., 0., 0.]]], dtype=float32)},

In [116]:
save_results = []
for myregion in myregions[0:3]:
    p, m = enformer_predict_on_sequence(enformer_model, myregion)
    save_results.append(save_batch_h5_prediction(p, m, output_dir, 'LuCaP_000'))


haplotype1: (17, 5313)
haplotype2: (17, 5313)
haplotype1: (17, 5313)
haplotype2: (17, 5313)
haplotype1: (17, 5313)
haplotype2: (17, 5313)


In [119]:
import pandas as pd

In [122]:
logfile = pd.DataFrame(save_results, columns=['motif', 'individual', 'status', 'sequence_source'])
logfile

motif individual     status sequence_source
0  chr1_103268936_103268945  LuCaP_000  completed             var
1  chr1_150333985_150333994  LuCaP_000  completed             var
2   chr12_45119102_45119111  LuCaP_000  completed             var

In [123]:
def check_queries(sample, queries, output_dir, logfile):
    """
    Check whether a given region, for an individual has been predicted and logged.

    Parameters:
        sample: str 
            The name/id of an individual
        queries: A batch
            A region in the genome in the form `chr_start_end`.
        output_dir: str (path)
            The folder where the predictions should have been logged. 
        logfile: pd.DataFrame 
            A dataframe of a log file or `None` if the log file does not exist. 
    
    Returns: dict
        'query': the query region if it has not been logged or predictions don't exist
        'logtype': whether it should be logged if it has not been logged

    If predictions exist and the query has been logged, this function returns None.
    """
    import pandas as pd
    import os
    import numpy as np

    if isinstance(logfile, type(None)):
        output = [{'query':query, 'logtype':'y'} for query in queries]
        return(output)

    elif isinstance(logfile, pd.DataFrame):# should have read it>
    print()

        id_logfile = logfile.loc[logfile['individual'] == sample, : ]
        # check if the file is saved
        if sequence_source == 'personalized': # prediction must be present in two folders
            queries_saved_h1 = [str(f'{output_dir}/{sample}/haplotype1/{query}_predictions.h5') for query in queries]
            queries_saved_h2 = [str(f'{output_dir}/{sample}/haplotype2/{query}_predictions.h5') for query in queries]
            queries_saved = np.array([os.path.isfile(q) for q in queries_saved_h1]) * np.array([os.path.isfile(q) for q in queries_saved_h2])
        elif sequence_source == 'reference':
            queries_saved = [str(f'{output_dir}/{sample}/haplotype0/{query}_predictions.h5') for query in queries]
            queries_saved = np.array([os.path.isfile(q) for q in queries_saved])
        # check if the file is in the logfile
        query_logged = np.array([(query in id_logfile.motif.values) for query in queries])
        # 
        if not query_logged.shape[0] == queries_saved.shape[0]:
            raise Exception("Lengths of queries logged and saved conditions are not the same")
        queries_condition = queries_saved * query_logged
        queries_condition = queries_condition.tolist()
        # 
        output = [{'query':queries[i], 'logtype':'n'} if qc is True else {'query':queries[i], 'logtype':'y'} for i, qc in enumerate(queries_condition)]

        return(output)

In [130]:
queries = [list(bb.values())[0] for bb in batch_region]
queries

['chr1_103268936_103268945',
 'chr1_150333985_150333994',
 'chr12_45119102_45119111',
 'chr8_146200116_146200125']

In [131]:
check_queries('LuCaP_000', queries, output_dir, logfile)

[{'query': 'chr1_103268936_103268945', 'logtype': 'n'},
 {'query': 'chr1_150333985_150333994', 'logtype': 'n'},
 {'query': 'chr12_45119102_45119111', 'logtype': 'n'},
 {'query': 'chr8_146200116_146200125', 'logtype': 'y'}]

In [127]:
batch_region

[{'query': 'chr1_103268936_103268945', 'logtype': 'y'},
 {'query': 'chr1_150333985_150333994', 'logtype': 'y'},
 {'query': 'chr12_45119102_45119111', 'logtype': 'y'},
 {'query': 'chr8_146200116_146200125', 'logtype': 'y'}]

In [47]:
save_batch_h5_prediction(pred, metadata, output_dir, 'LuCaP_000')

['chr12_45119102_45119111', 'LuCaP_000', 'completed', 'ref']

One function to do it all

In [49]:
pred['haplotype0'].shape

(1, 17, 5313)

In [26]:
sequence_source = 'personalized'

In [29]:
def collect_inputs_into_dict(batch_region):

    output_sequences = [utilities.create_input_for_enformer(region_details=each_region, vcf_func=myvcf, fasta_func=fasta_object, hap_type = 'both', resize_for_enformer=True, resize_length=None, sequence_source='personalized') for each_region in batch_region] #>> [{sequence:{}, region: '', sequence_source: ''}, ..., {...}]
    #output_metadata = [{'logtype': each_region['logtype']} for each_region in batch_region]
    output = [out for out in output_sequences if out is not None]

    if sequence_source == 'personalized':
        hap1 = [out_dict['sequence']['haplotype1'] for out_dict in output]
        hap2 = [out_dict['sequence']['haplotype2'] for out_dict in output]

        variant_details = [{'sequence_source': out_dict['sequence_source'],
                            'region': out_dict['region'],
                            'logtype': out_dict['logtype']} for out_dict in output]
        
        tensor1 = {'haplotype1': tf.concat([utilities.create_enformer_input(s) for s in hap1], axis=0)}
        tensor2 = {'haplotype2': tf.concat([utilities.create_enformer_input(s) for s in hap2], axis=0)}
        #query_names = [out_dict['query'] for out_dict in output]
    elif sequence_source == 'reference':
        hap0 = [out_dict['sequence']['haplotype0'] for out_dict in output]
        variant_details = [{'sequence_source': out_dict['sequence_source'],
                            'region': out_dict['region'],
                            'logtype': out_dict['logtype']} for out_dict in output]

        tensor1 = {'haplotype0': tf.concat([utilities.create_enformer_input(s) for s in hap0], axis=0)}
        tensor2 = None #{'haplotype2': tf.concat([create_enformer_input(s) for s in hap2], axis=0)}

    return( tensor1, tensor2, variant_details )


In [30]:
import time

tic = time.perf_counter()
t1, t2, pc = collect_inputs_into_dict(batch_region)
toc = time.perf_counter()

{'query': 'chr1_103268936_103268945', 'logtype': 'n'}
(1, 393216, 4)
(1, 393216, 4)
{'query': 'chr1_150333985_150333994', 'logtype': 'n'}
(1, 393216, 4)
(1, 393216, 4)
{'query': 'chr12_45119102_45119111', 'logtype': 'n'}
(1, 393216, 4)
(1, 393216, 4)
{'query': 'chr8_146200116_146200125', 'logtype': 'n'}


AttributeError: module 'utilities' has no attribute 'create_enformer_input'

In [35]:
predict_s1 = enformer_model.predict_on_batch(t1['haplotype1'])['human'].numpy()[: , range(448 - 8, (448 + 8 + 1)), : ]
predict_s2 = enformer_model.predict_on_batch(t2['haplotype2'])['human'].numpy()[: , range(448 - 8, (448 + 8 + 1)), : ]

2023-01-31 17:59:18.155816: I tensorflow/stream_executor/cuda/cuda_blas.cc:1786] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once.
2023-01-31 17:59:19.101483: I tensorflow/stream_executor/cuda/cuda_dnn.cc:384] Loaded cuDNN version 8100


In [72]:
predict_s0 = enformer_model.predict_on_batch(t1['haplotype0'])['human'].numpy()[: , range(448 - 8, (448 + 8 + 1)), : ]


In [4]:
import subprocess
subprocess.call('nvidia-smi')

Failed to initialize NVML: Insufficient Permissions


4

In [73]:
pc

[{'sequence_source': 'ref',
  'region': 'chr1_103268936_103268945',
  'logtype': 'n'},
 {'sequence_source': 'ref',
  'region': 'chr1_150333985_150333994',
  'logtype': 'n'},
 {'sequence_source': 'ref',
  'region': 'chr12_45119102_45119111',
  'logtype': 'n'}]

In [34]:
def save_batch_h5_prediction(batch_predictions, metadata, output_dir, sample):

    import h5py

    #output = []

    for key, values in batch_predictions.items():
        houtput = os.path.join(output_dir, key)
        if not os.path.exists(houtput): os.makedirs(houtput)
        for i in range(0, values.shape[0]):
            region = metadata[i]['region']
            h5save = str(f'{houtput}/{region}_predictions.h5')
            with h5py.File(h5save, 'w') as hf:
                hf.create_dataset(region, data=values[i, :])
            
    output = [[md['region'], sample, 'completed', md['sequence_source']] for md in metadata if md['logtype'] == 'y']

    return(output)


In [76]:
bpredictions = {'haplotype0': predict_s0}

In [79]:
save_batch_h5_prediction(bpredictions, pc, sample='', output_dir=output_dir)

# save_batch_h5_prediction(predict_s2, pc, sample='', output_dir=output_dir, haplotype='haplotype2')

[]

In [69]:
pc

[{'sequence_source': 'var',
  'region': 'chr1_103268936_103268945',
  'logtype': 'n'},
 {'sequence_source': 'var',
  'region': 'chr1_150333985_150333994',
  'logtype': 'n'},
 {'sequence_source': 'var',
  'region': 'chr12_45119102_45119111',
  'logtype': 'n'}]

In [51]:
predict_s['human'].numpy()[: , range(448 - 8, (448 + 8 + 1)), : ][0, :].shape

(17, 5313)

In [40]:
type(predict_s)

dict

In [51]:
predict_s['human'].shape

TensorShape([2, 896, 5313])

In [1]:
def batch(iterable, n=1):
    l = len(iterable)
    for ndx in range(0, l, n):
        yield iterable[ndx:min(ndx + n, l)]

In [7]:
mylist = list(range(0, 5))
mylist


[0, 1, 2, 3, 4]

In [10]:
list(batch(mylist, 8))

[[0, 1, 2, 3, 4]]

In [2]:
import h5py
import numpy as np

In [6]:
dir = './../prediction_folder/enformer_predictions/geuvadis_AR/predictions_2023-02-11'
motif = 'chrX_87537332_87537341'

In [7]:
output = {}
fle = f'{dir}/{motif}_predictions.h5'
if os.path.isfile(fle):
    with h5py.File(fle, 'r') as f:
        filekey = list(f.keys())[0]
        output[motif] = np.vstack(list(f[filekey]))

In [4]:
output['chr10_87771654_87771663'].shape

(5313, 1)

In [9]:
output['chrX_87537332_87537341'].shape

(17, 5313)

## cProfiler

In [10]:
import cProfile, pstats
from pstats import SortKey

In [11]:
p = pstats.Stats('../../../../profile.out')
p.sort_stats('cumtime').print_stats()

Wed Feb  1 04:46:24 2023    ../../../../profile.out

         116211218 function calls (106625407 primitive calls) in 1053.737 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    936/1    0.007    0.000 1053.738 1053.738 {built-in method builtins.exec}
        1    0.000    0.000 1053.738 1053.738 /lus/grand/projects/covid-ct/imlab/users/temi/projects/TFXcan/TFPred_pipeline/scripts/enformer_predict/new_enformer_predict.py:1(<module>)
        1    0.005    0.005 1052.274 1052.274 /lus/grand/projects/covid-ct/imlab/users/temi/projects/TFXcan/TFPred_pipeline/scripts/enformer_predict/new_enformer_predict.py:28(main)
       14    0.000    0.000 1014.096   72.435 /home/temi/miniconda3/envs/dl-tools/lib/python3.10/threading.py:288(wait)
       38 1014.095   26.687 1014.095   26.687 {method 'acquire' of '_thread.lock' objects}
        1    0.000    0.000 1014.095 1014.095 /lus/grand/projects/covid-ct/imlab/users/temi/projects/TFX

import itertools

In [7]:
import itertools
import multiprocessing
import os

In [8]:
os.chdir('/lus/grand/projects/covid-ct/imlab/users/temi/projects/TFXcan/TFPred_pipeline/test_run')
os.getcwd()

'/lus/grand/projects/covid-ct/imlab/users/temi/projects/TFXcan/TFPred_pipeline/test_run'

In [19]:
predictions_path = './../prediction_folder/enformer_predictions/geuvadis_AR/predictions_2023-02-01'
individual_predictions = os.listdir(os.path.join(predictions_path))
individual_predictions[1:5]

['HG00239', 'HG00128', 'HG00247', 'HG00311']

In [21]:
fdir = os.path.join(predictions_path, 'HG00239', 'haplotype1')

In [24]:
len(os.listdir(fdir))

18654

In [22]:
sum(os.path.getsize(f) for f in os.listdir(fdir) if os.path.isfile(f))

0

In [25]:
empty = []
for ip in individual_predictions:
    fdir = os.path.join(predictions_path, ip, 'haplotype1')
    if len(os.listdir(fdir)) != 18654:
        empty.append(ip)

empty
    

['HG00247', 'HG00124']

In [12]:
motifs = os.listdir(os.path.join(predictions_path, 'haplotype1'))

In [17]:
os.listdir(os.path.join(predictions_path, 'haplotype2'))

[]

In [13]:
motifs[1:5]

[]

In [4]:
motifs = ['chrX_87537332_87537341','chr15_72707586_72707595',
                'chr2_238062837_238062846',
                'chr2_236926577_236926586',
                'chr2_146496463_146496472',
                'chr15_87147461_87147470',
                'chr2_27719423_27719432',
                'chrX_36834447_36834456',
                'chr2_28929008_28929017',
                'chr15_69693730_69693739',
                'chr2_129537996_129538005',
                'chr15_65250515_65250524',
                'chr2_35172176_35172185',
                'chr2_224231575_224231584',
                'chr15_30980323_30980332',
                'chr15_79338374_79338383',
                'chr2_66048726_66048735']
motifs

['chrX_87537332_87537341',
 'chr15_72707586_72707595',
 'chr2_238062837_238062846',
 'chr2_236926577_236926586',
 'chr2_146496463_146496472',
 'chr15_87147461_87147470',
 'chr2_27719423_27719432',
 'chrX_36834447_36834456',
 'chr2_28929008_28929017',
 'chr15_69693730_69693739',
 'chr2_129537996_129538005',
 'chr15_65250515_65250524',
 'chr2_35172176_35172185',
 'chr2_224231575_224231584',
 'chr15_30980323_30980332',
 'chr15_79338374_79338383',
 'chr2_66048726_66048735']

In [15]:
global read_file
def read_file(motif, dir, haplotype):
    import os
    import h5py
    import numpy as np

    output = {}
    fle = os.path.join(dir, haplotype, f'{motif}_predictions.h5')  #f'{dir}//{motif}_predictions.h5'
    if os.path.isfile(fle):
        with h5py.File(fle, 'r') as f:
            filekey = list(f.keys())[0]
            output[motif] = np.vstack(list(f[filekey]))
    else:
        print(f'[ERROR] {motif} predictions file does not exist.')

    return(output)

In [20]:
pool = multiprocessing.Pool(16)
outputs_list_1 = pool.starmap(read_file, itertools.product(motifs, [predictions_path], ['haplotype1']))
outputs_list_2 = pool.starmap(read_file, itertools.product(motifs, [predictions_path], ['haplotype2']))

predictions_1 =  {k: v for d in outputs_list_1 for k, v in d.items()}
predictions_2 =  {k: v for d in outputs_list_2 for k, v in d.items()}

Process ForkPoolWorker-63:
Process ForkPoolWorker-54:
Process ForkPoolWorker-60:
Process ForkPoolWorker-55:
Process ForkPoolWorker-56:
Process ForkPoolWorker-61:
Process ForkPoolWorker-58:
Process ForkPoolWorker-50:
Process ForkPoolWorker-59:
Process ForkPoolWorker-49:
Process ForkPoolWorker-62:
Process ForkPoolWorker-64:
Process ForkPoolWorker-51:
Process ForkPoolWorker-53:
Process ForkPoolWorker-57:
Process ForkPoolWorker-52:
Traceback (most recent call last):
Traceback (most recent call last):
  File "/home/temi/miniconda3/envs/dl-tools/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/home/temi/miniconda3/envs/dl-tools/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/home/temi/miniconda3/envs/dl-tools/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
Traceback (most recent call last):
  File "/home/temi/miniconda3/envs/dl-tools/lib/python3.10

In [26]:
predictions_1['chrX_36834447_36834456'].shape, predictions_2['chrX_36834447_36834456'].shape

((17, 5313), (17, 5313))

In [28]:
sum_predictions = predictions_1['chrX_36834447_36834456'] + predictions_2['chrX_36834447_36834456']
sum_predictions.shape

(17, 5313)

In [ ]:
sorted(list(predictions_1['haplotype1'].keys())) == sorted(list(pooled_dictionary[haplotypes[1]].keys()))

In [31]:
sorted(list(predictions_1.keys())) == sorted(list(predictions_2.keys()))

True

In [41]:
newline = '\n'

f"[ERROR] Fatal: Haplotypes 1 and 2 regions are different. Haplotype1 length is {len(sorted(list(predictions_1.keys())))} and Haplotype2 length is {len(sorted(list(predictions_2.keys())))}"

'[ERROR] Fatal: Haplotypes 1 and 2 regions are different. Haplotype1 length is 17 and Haplotype2 length is 17'

In [6]:
import tensorflow as tf
tf.config.experimental.list_physical_devices()

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'),
 PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU'),
 PhysicalDevice(name='/physical_device:GPU:1', device_type='GPU'),
 PhysicalDevice(name='/physical_device:GPU:2', device_type='GPU'),
 PhysicalDevice(name='/physical_device:GPU:3', device_type='GPU')]

In [8]:
predictions_path

'./../prediction_folder/enformer_predictions/freedman_FOXA1/predictions_2023-01-11/LuCaP_173'

In [11]:
list(itertools.product(motifs, [predictions_path], ['haplotype1']))

[('chrX_87537332_87537341',
  './../prediction_folder/enformer_predictions/freedman_FOXA1/predictions_2023-01-11/LuCaP_173',
  'haplotype1'),
 ('chr15_72707586_72707595',
  './../prediction_folder/enformer_predictions/freedman_FOXA1/predictions_2023-01-11/LuCaP_173',
  'haplotype1'),
 ('chr2_238062837_238062846',
  './../prediction_folder/enformer_predictions/freedman_FOXA1/predictions_2023-01-11/LuCaP_173',
  'haplotype1'),
 ('chr2_236926577_236926586',
  './../prediction_folder/enformer_predictions/freedman_FOXA1/predictions_2023-01-11/LuCaP_173',
  'haplotype1'),
 ('chr2_146496463_146496472',
  './../prediction_folder/enformer_predictions/freedman_FOXA1/predictions_2023-01-11/LuCaP_173',
  'haplotype1'),
 ('chr15_87147461_87147470',
  './../prediction_folder/enformer_predictions/freedman_FOXA1/predictions_2023-01-11/LuCaP_173',
  'haplotype1'),
 ('chr2_27719423_27719432',
  './../prediction_folder/enformer_predictions/freedman_FOXA1/predictions_2023-01-11/LuCaP_173',
  'haplotype1')

In [4]:
list(itertools.product(['a', 'c'], ['b1', 'b2']))

[('a', 'b1'), ('a', 'b2'), ('c', 'b1'), ('c', 'b2')]